packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.internal.js

```diff
@@ -3263,4 +3263,45 @@ describe('ReactHooksWithNoopRenderer', () => {
     });
     expect(ReactNoop).toMatchRenderedOutput('ABC');
   });
+
+  it('keeps intermediate state updates (issue #18497)', () => {
+    let _dispatch;
+    function Counter() {
+      const [list, dispatch] = React.useReducer((l, c) => l.concat([c]), []);
+      _dispatch = dispatch;
+
+      const json = JSON.stringify(list);
+      Scheduler.unstable_yieldValue('Render ' + json);
+      useLayoutEffect(() => {
+        Scheduler.unstable_yieldValue('Commit ' + json);
+      });
+
+      return json;
+    }
+
+    act(() => {
+      ReactNoop.render(<Counter />);
+      expect(Scheduler).toFlushAndYieldThrough(['Render []', 'Commit []']);
+      expect(ReactNoop).toMatchRenderedOutput('[]');
+    });
+
+    act(() => {
+      _dispatch(1);
+      expect(Scheduler).toFlushAndYieldThrough(['Render [1]']);
+
+      _dispatch(2);
+      expect(Scheduler).toFlushAndYieldThrough(['Commit [1]']);
```

Comment:
Not sure if this is expected either. Should the dispatch prevent this one from being committed? I can't remember how this is supposed to work.